### DISCUSSION TITLE
3D multi scale tesselation

### DISCUSSION THREAD

**Discussion by ravikiranbollineni** [neperfepx/neper#353] [Q&A] [answered]
Hello Romain,

I am able to create a 2D multi scale tessellation using the code below:
```
neper -T -n 50::from_morpho -dim 2 -domain "square(0.3,0.3)" -morpho "diameq:1,1-sphericity:lognormal(0.145,0.03)::lamellar(w=0.005)" -reg 1:1 -statface diameq -morphooptistop "nlopt_reps=1e-6" -o n50-id11
```

The geometry looks fine for my application

![img3](https://user-images.githubusercontent.com/82678406/147128788-6fc9e39c-7f72-4db5-91ab-108ae4fefc23.png)

But when I am creating  3D model with a thickness of 0.001 using the code below  :
```
neper -T -n 50::from_morpho -dim 3 -domain "cube(0.3,0.3,0.005)" -morpho "diameq:1,1-sphericity:lognormal(0.145,0.03)::lamellar(w=0.005)" -reg 1:1 -statface diameq -morphooptistop "nlopt_reps=1e-6" -o n50-id13
```

[
![img9](https://user-images.githubusercontent.com/82678406/147136345-dad8bfeb-b751-44f7-b5bb-3063cae67700.png)
](url)

I am not getting exactly the same geometry as above with thickness 0.005. 
can you help in this regards?

Thank you,
Ravi.

**Comment by dikshanshs** [comment]
@ravikiranbollineni : For 3D, seed positions are different in comparison of 2D

**Reply by ravikiranbollineni** [reply]
Thanks @dikshanshs for answering question. Can you explain the process to get the 3D model that looks similar to the 2D model as above with thickness 0.005.

**Reply by dikshanshs** [reply]
@ravikiranbollineni you can use seed position and use columnar or try the following command
`neper -T -n 50::from_morpho -dim 3 -domain "cube(0.3,0.3,0.005)" -morpho "columnar(z)::lamellar(w=0.003)" -reg 1:1 -morphooptistop "nlopt_reps=1e-6" -o n50-id13`

**Comment by rquey** [comment, marked as answer]
There are several problems in your command:

- the domain is very thin compared to the average in-plane cell size, and so the `1-sphericity` distribution cannot be matched (cells tend to have much smaller sphericities than imposed). `diameq:1` is imposing equal-size cells, but this would require small sphericities. The structure you get is a (bad) compromise between these contradictory inputs. It is often possible to "detect" bad/contradictory/impossible inputs from the value of the objective function, which should rapidly become small with respect to 1 (as in your 2D case).

- In 2D, the lamella normals are in plane (x-y), while in 3D, they are 3D.  This is why the 3D tessellation does not have in-plane lamella normals.

I assume that what you want is an extrusion of the 2D tessellation. In this case, the best probably is to proceed in 2D steps:

1. Generate a 2D structure (1st scale):

```
neper -T -n 50 -dim 2 -domain "square(0.3,0.3)" -morpho "diameq:1,1-sphericity:lognormal(0.145,0.03)" -o 2d
```

2. Generate a 3D structure from the 2D structure, and add the 2nd scale:

```
neper -T -n 50::from_morpho -domain "cube(0.3,0.3,0.005)" -morpho "voronoi::lamellar(w=0.003,v=random(x,y))" -morphooptiini "file(2d.tess)" -o 3d
```

Visualize:
```
neper -V 2d.tess -cameracoo 1:1:0.5 -cameralookat 0.15:0.15:0 -cameraangle 20 -cameraprojection perspective -camerasky 0:0:1 -dataedgerad 0.0003 -datacellcol id -print 2d
neper -V 3d.tess -cameracoo 1:1:0.5 -cameralookat 0.15:0.15:0 -cameraangle 20 -cameraprojection perspective -camerasky 0:0:1 -dataedgerad 0.0003 -datacellcol "scaleid(1)" -print 3d
```
![image](https://user-images.githubusercontent.com/2659376/148358186-77cdac79-5949-45a6-b877-34444b81b03a.png)
![image](https://user-images.githubusercontent.com/2659376/148358199-9529f885-7c7f-498a-a703-0582db84b140.png)


Needs version 4.2.1-21.

**Reply by ravikiranbollineni** [reply]
Thanks for answering Romain. I am not able to find the Neper 4.2.1-21 version. Could you let me know where I can find 4.2.1 version.
Thanks.

**Reply by rquey** [reply]
It is the version you can get here: https://github.com/rquey/neper.

**Reply by ravikiranbollineni** [reply]
Thanks Romain for the help.

**Comment by xshang93** [comment]
@rquey Came across this post and I'm curious: is there a way to extrude 2d to 3d when doing a 3 scale tessellation? I didn't have any success because the seeds loaded will always be overwritten on the 2nd scale tessellation.
